pkg/plugins/runtime/gateway/generator.go

```diff
@@ -414,12 +424,18 @@ func RedistributeWildcardRoutes(
 				host.Routes = nil
 			}
 
-			// Note that if we already have a virtualhost for this
-			// name, and add the route to it, it might be a duplicate.
-			host.Routes = append(host.Routes, r)
-			host.Hostname = n
-
-			hostsByName[n] = host
+			appendRoutes(n, host)
+
+			if suffix := strings.TrimPrefix(n, "*."); len(suffix) != len(n) {
+				// An alternative to this might be precalculating
+				// a table of suffixes to hosts
+				for hostName, host := range hostsByName {
+					if !strings.HasSuffix(hostName, suffix) {
+						continue
+					}
+					appendRoutes(hostName, host)
```

Comment:
Might be just me but I find the function just making things harder to read.

Would the alternative mentioned here be simpler?

```go
prefixedRoutes := map[string]GatewayHost{}
for n, host := range hostsByName {
   if suffix := strings.TrimPrefix(n, "*."); len(suffix) != len(n) {
       prefixRoutes[suffix] := host
   }
}
for _, n := range names {
    // ....
    // Note that if we already have a virtualhost for this
   // name, and add the route to it, it might be a duplicate.
  host.Routes = append(host.Routes, r)
  host.Hostname = n
  hostsByName[n] = host
  for suffix, prefixedHost := range prefixedRoutes {
     if n != suffix && strings.HasSuffix(n, suffix) {
         host.Routes = append(host.Routes, prefixedHost.Routes)
     }
  }
}
```

Reply:
I think this changes the semantics. I check the hostnames of wildcard routes for `*.` prefixed strings and then append the wildcard routes `r` to `hostsByName[name]` routes if `name` matches.

I think yours checks `hostsByName[name]` for if `name` is `*.` prefixed and then appends the routes from `hostsByName[name]` to the wildcard routes.

I think the structure of `GatewayHost` is somewhat confusing. We should also consider whether `hostname` on a `Listener` should allow `*.` but this PR only allows for `*.` in `Route` `hostname`.

Reply:
@lahabana I improved the commit somewhat, I'm resolving this for now to merge this, since it lets us pass the Gateway API conformance tests (more or less, see #3450)